### ISSUE TITLE
Bug: wrong x and y in lat_lon_to_closest_variable_indices and lat_lon_boundingbox_to_variable_indices

### ISSUE THREAD

**Issue by SarahAlidoost** [eWaterCycle/ewatercycle#104]
The longitude value is obtained by **x** coordinates whereas latitude by **y**:
```python
"longitude": bmi.get_grid_x(grid),
"latitude": bmi.get_grid_y(grid),
```

This is not the case in two functions implemented in util.py:
[lat_lon_to_closest_variable_indices](https://github.com/eWaterCycle/ewatercycle/blob/1156a4ff7cdc1fc7143d0cc1137695c405581f55/ewatercycle/util.py#L43)
[lat_lon_boundingbox_to_variable_indices](https://github.com/eWaterCycle/ewatercycle/blob/1156a4ff7cdc1fc7143d0cc1137695c405581f55/ewatercycle/util.py#L69)

**Comment by Peter9192** [comment]
I think we've seen this before. Could it be that this was originally coded like that for wflow (or perhaps it was pcrglob)? I think one of these models might have reversed this convention internally.

**Comment by SarahAlidoost** [comment]
yes, I am exploring this now. Here the documentation of [grpc4bmi for these functions](https://grpc4bmi.readthedocs.io/en/0.2.x/api/grpc4bmi.bmi_grpc_client.html#grpc4bmi.bmi_grpc_client.BmiClient.get_grid_x) that is not clear to me what is meant by **streamwise and transverse** directions.  I see that it is reversed in the [pcrglob comparison notebook ](https://github.com/eWaterCycle/era5-comparison/blob/master/pcrglobwb/03_pcrglobwb_fullruns.ipynb), and [wflow notebook](https://github.com/eWaterCycle/era5-comparison/blob/master/wflow/03_wflow_fullruns.ipynb).  But [lisflood](https://github.com/eWaterCycle/era5-comparison/blob/master/lisflood/03_lisflood_fullruns.ipynb) uses the correct notations, y for latitude and x for longitude. Not confusing at all.

**Comment by SarahAlidoost** [comment]
In `wflow bmi.py`, the function [get_grid_x](https://github.com/eWaterCycle/wflow/blob/fdb404c543c199b942a0970411301e2c3a5cad84/wflow/bmi.py#L374) is defined based on column centers.

**Comment by Peter9192** [comment]
> 
> 
> In `wflow bmi.py`, the function [get_grid_x](https://github.com/eWaterCycle/wflow/blob/fdb404c543c199b942a0970411301e2c3a5cad84/wflow/bmi.py#L374) is defined based on column centers.

Your link directs to the abstractmethod, the actual implementation is [here](https://github.com/eWaterCycle/wflow/blob/fdb404c543c199b942a0970411301e2c3a5cad84/wflow/wflow_bmi.py#L1264). Indeed, `get_grid_x` points to wflows `wf_supplyMapYAsNumpy` and vice versa. Not sure if that's intentional.

**Comment by Peter9192** [comment]
> 
> 
> yes, I am exploring this now. Here the documentation of [grpc4bmi for these functions](https://grpc4bmi.readthedocs.io/en/0.2.x/api/grpc4bmi.bmi_grpc_client.html#grpc4bmi.bmi_grpc_client.BmiClient.get_grid_x) that is not clear to me what is meant by **streamwise and transverse** directions. I see that it is reversed in the [pcrglob comparison notebook ](https://github.com/eWaterCycle/era5-comparison/blob/master/pcrglobwb/03_pcrglobwb_fullruns.ipynb), and [wflow notebook](https://github.com/eWaterCycle/era5-comparison/blob/master/wflow/03_wflow_fullruns.ipynb). But [lisflood](https://github.com/eWaterCycle/era5-comparison/blob/master/lisflood/03_lisflood_fullruns.ipynb) uses the correct notations, y for latitude and x for longitude. Not confusing at all.

The BMI spec for `get_grid_x` is [here](https://bmi-spec.readthedocs.io/en/latest/#get-grid-x), it says x is for the "first coordinate direction"  and y is for the "second coordinate direction". Not sure if that helps. 

Streamwise and transverse sound to me as directions relative to the flow direction, I don't think that makes a lot of sense for the types of models we work with.

**Comment by Peter9192** [comment]
Looking at the source code for pcraster (i.e. what wflow uses), the only useful reference I can find is here:

https://github.com/pcraster/pcraster/blob/750c2c3ae5a4e2292391e6bab530050f7d6a0c59/source/pcraster_python/PCRaster.cc#L616

which seems to imply that x is indeed "west-east" (i.e. longitude) and y is "north-south".

**Comment by Peter9192** [comment]
In the latest version of wflow (and also in the abandoned master branch of the python implementation), get_grid_x refers to the x-coordinate and the longitude, as one would expect:

Julia (latest model version): 
[get_grid_x](https://github.com/Deltares/Wflow.jl/blob/856f713f6dd71d5849f12978d907d7a572709a8f/src/bmi.jl#L255)
[read_x_axis](https://github.com/Deltares/Wflow.jl/blob/1240d69723717400ef2c4238640ded0efe989f6e/src/io.jl#L1275)

Python (most recent old version):
[get_grid_x](https://github.com/openstreams/wflow/blob/ad0934ed53519d7ef5793cdd12b72201dd979720/wflow/wflow_bmi.py#L1246)

I suspect the switch from x to y was accidentily introduced [here](https://github.com/openstreams/wflow/commit/bef9e53e3a9ec95f1e1e6ad354c59e1b36f627e1), but I cannot trace it far enough to confirm it was indeed accidental or intentional.

In terms of what to do now, I see a few options

- Fix it in the ewatercycle branch of wflow. This would require updating the docker container etc, and we're already using a branch with fixes.. not great!
- Eventually it would be nice to switch to a more recent version of wflow, bmi, etc. Postpone until then.
- (Temporarily) implement a workaround in the ewatercycle package clearly pointing to this bug report and discussion.

**Comment by SarahAlidoost** [comment]
> In the latest version of wflow (and also in the abandoned master branch of the python implementation), get_grid_x refers to the x-coordinate and the longitude, as one would expect:
> 
> Julia (latest model version):
> [get_grid_x](https://github.com/Deltares/Wflow.jl/blob/856f713f6dd71d5849f12978d907d7a572709a8f/src/bmi.jl#L255)
> [read_x_axis](https://github.com/Deltares/Wflow.jl/blob/1240d69723717400ef2c4238640ded0efe989f6e/src/io.jl#L1275)
> 
> Python (most recent old version):
> [get_grid_x](https://github.com/openstreams/wflow/blob/ad0934ed53519d7ef5793cdd12b72201dd979720/wflow/wflow_bmi.py#L1246)
> 
> I suspect the switch from x to y was accidentily introduced [here](https://github.com/openstreams/wflow/commit/bef9e53e3a9ec95f1e1e6ad354c59e1b36f627e1), but I cannot trace it far enough to confirm it was indeed accidental or intentional.

Thank you for exploring the issue further. Right, the same here, I couldn't find more information there. Just one point that the current wflow bmi is available on forked wflow repository in [eWaterCycle branch docker-fixes-2020.1.1](https://github.com/eWaterCycle/wflow/blob/docker-fixes-2020.1.1/wflow/bmi.py). However, the codes related to `get_grid_x` is the same as those in the link that you shared. 
> 
> In terms of what to do now, I see a few options
> 
> * Fix it in the ewatercycle branch of wflow. This would require updating the docker container etc, and we're already using a branch with fixes.. not great!
> * Eventually it would be nice to switch to a more recent version of wflow, bmi, etc. Postpone until then.
> * (Temporarily) implement a workaround in the ewatercycle package clearly pointing to this bug report and discussion.

I like adding two more items:
- exploring this issue for pcrglob, too. 
- removing/refactoring `lat_lon_to_closest_variable_indices` and `lat_lon_boundingbox_to_variable_indices`  in `util.py`. Because the implementations are not the same for all models.

**Comment by SarahAlidoost** [comment]
> > In `wflow bmi.py`, the function [get_grid_x](https://github.com/eWaterCycle/wflow/blob/fdb404c543c199b942a0970411301e2c3a5cad84/wflow/bmi.py#L374) is defined based on column centers.
> 
> Your link directs to the abstractmethod, the actual implementation is [here](https://github.com/eWaterCycle/wflow/blob/fdb404c543c199b942a0970411301e2c3a5cad84/wflow/wflow_bmi.py#L1264). Indeed, `get_grid_x` points to wflows `wf_supplyMapYAsNumpy` and vice versa. Not sure if that's intentional.

Great, thanks for this. So if coordinates are reversed here, I am not sure why we reverse them again either in the comparison or util functions.

**Comment by Peter9192** [comment]
> Great, thanks for this. So if coordinates are reversed here, I am not sure why we reverse them again either in the comparison or util functions.

I think the first reversal is an error in the BMI implementation, and reversing them again is correcting that error.

**Comment by Peter9192** [comment]
Is this what we're using for PCRGlob? https://github.com/eWaterCycle/PCR-GLOBWB_model/blob/e25a07bf63dae0237a7e7039454e96151f9d41c1/model/bmiPcrglobwb.py#L569

There it seems to be reversed as well.

**Comment by SarahAlidoost** [comment]
> Is this what we're using for PCRGlob? https://github.com/eWaterCycle/PCR-GLOBWB_model/blob/e25a07bf63dae0237a7e7039454e96151f9d41c1/model/bmiPcrglobwb.py#L569
> 
> There it seems to be reversed as well.

I don't know the pcrglob bmi source, it is not documented in its comparison readme yet. OK, I agree with you that this is an error in both the BMI implementations of wflow and pcrglob.

**Comment by SarahAlidoost** [comment]
oh, I found the pcrglob source https://github.com/eWaterCycle/PCR-GLOBWB_model/tree/bmi_fixes, it is mentioned [here](https://github.com/eWaterCycle/grpc4bmi-examples/blob/master/pcrglob/Dockerfile).

**Comment by Peter9192** [comment]
Yes, although we're currently using this docker image: "ewatercycle/pcrg-grpc4bmi:setters". In both cases the `get_grid_x` function refers to latitudes instead of longitudes.

**Comment by SarahAlidoost** [comment]
Conclusion: 
- (Temporarily) implement a workaround in the ewatercycle package clearly pointing to this bug report and discussion.
- Fix it in the bmi implementations of wflow and pcrglob later. (we will make a work item for that).

**Comment by Peter9192** [comment]
After working on #140, I've reduced the difference between our models to:

Lisflood: 
- `get_grid_x` --> longitude
- `get_grid_y` --> latitude
- `get_grid_shape` --> by testing we know this returns (len(y), len(x) => (nlat, nlon)

Wflow and PCRGlob:
- `get_grid_x` --> latitude
- `get_grid_y` --> longitude
- `get_grid_shape` --> by testing we know this returns (len(x), len(y) => (nlat, nlon)

Notice how the shape is always (nlat, nlon). But there seem to be two conflics with intuition:

1. Intuitively, x is longitude and y latitude, for both x and longitude are usually the "horizontal" coordinate.
2. Intuitively, the shape of a grid is expected to be (nx, ny). This idea is reinforced the [BMI docstrings](https://bmi-spec.readthedocs.io/en/latest/#get-grid-x) for `get_grid_x` and `get_grid_y`, which refer to the *first* and *second* coordinate direction, respectively.

However, the [BMI documentation](https://bmi-spec.readthedocs.io/en/latest/model_grids.html#structured-grids) explains that that the BMI grid order uses "ij" indexing:

> In the BMI, dimensional information is ordered with “ij” indexing (as opposed to “xy”). For example, for the uniform rectilinear grid shown below, the get_grid_shape function would return the array [4, 5]. **If there was a third dimension, its length would be listed first**.

Importantly, the **first** dimension is listed **last** in the shape. This resolves conflict number 2.

Furthermore, [here](https://bmi-spec.readthedocs.io/en/latest/model_grids.html#rectilinear) they explain that:

> The get_grid_y function provides an array (whose length is the number of rows) that gives the y-coordinate for each row.
> The get_grid_x function provides an array (whose length is the number of columns) that gives the x-coordinate for each column.

Their example grid looks roughly like this:
``` 
 * -- * -- * -- * -- *
 |    |    |    |    |        number of column nodes: 5  --> lenght of x-coordinate: 5
 * -- * -- * -- *  --*        number of row nodes: 4 --> lenght of y-coordinate: 4
 |    |    |    |    |        
 * -- * -- * -- * -- *
 |    |    |    |    |        shape: (4, 5) --> (nrows, ncols) --> (ny, nx) 
 * -- * -- * -- *  --* 
```

This seems to be consistent with the implementation of lisflood. With respect to Wflow and PCRGlob, it seems in both BMI implementations, `get_grid_shape` does map to the internal representation (nrows, ncols). However, `get_grid_x` is explicitly mapped to latitude and `y` to longitude. If we would swap those, we would get: 

- `get_grid_x` --> longitude => len(x) would be nlon
- `get_grid_y` --> latitude => len(y) would be nlat
- `get_grid_shape` --> (nlat, nlon) => (len(y), len(x))

so everything would be consistent. 

Conclusions: 
- x is longitude, 
- y is latitude, 
- shape is (ny, nx)
- Need to swap x and y in PCRGlob and Wflow BMI. 
- Lisflood is fine.

-------------------
PS: I was having fun with numpy:

To reconstruct this grid with numpy, we could do something like this:

```py
In [1]: import numpy as np
   ...:
   ...: lon = [1, 2, 3, 4, 5]
   ...: lat = [1, 2, 3, 4]
   ...: y, x = np.meshgrid(lat, lon, indexing='ij')

In [2]: x
Out[2]:
array([[1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5]])

In [3]: y
Out[3]:
array([[1, 1, 1, 1, 1],
       [2, 2, 2, 2, 2],
       [3, 3, 3, 3, 3],
       [4, 4, 4, 4, 4]])

In [4]: x.shape
Out[4]: (4, 5)
```

Interestingly, the exact same result is obtained with (default) `'xy'` indexing, if `x` or `lon` comes first in the construction:

```python
In [1]: import numpy as np
   ...:
   ...: lon = [1, 2, 3, 4, 5]
   ...: lat = [1, 2, 3, 4]

In [2]: x, y = np.meshgrid(lon, lat)

In [3]: x
Out[3]:
array([[1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5]])

In [4]: y
Out[4]:
array([[1, 1, 1, 1, 1],
       [2, 2, 2, 2, 2],
       [3, 3, 3, 3, 3],
       [4, 4, 4, 4, 4]])

In [5]: x.shape
Out[5]: (4, 5)
```
Notice that the shape is again `(ny, nx)`.
